### PR TITLE
Support maps in gen_mod_deps

### PR DESCRIPTION
**Pull request by chrzaszcz**
Prepare module deps for module opts in maps - now modules using maps can be deps of other modules.
- Update dependency processing and resolution.
- Update dependency format - options are mandatory as both `[]` and `#{}` are possible now. This does not seem to be worse in any way, so I think it is a permanent change.
- Update type specs, improve type reuse.
- Update small tests to check the opts in maps. The transitional phase should be short and I see no reason to keep both formats in tests - it would be complicated.

Minor changes:
- Use host types.
- Call module options `opts`, not `args` or `params`.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/108228) / small_tests / e4b60ba0458309df452e49d3f7a2f5cbde368d6c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3555/108228/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3555/108228/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-22_10.17.18/lib.mongooseim.logs/run.2022-02-22_10.17.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/108270) / small_tests / 39b89063306a2e4caeb2d920dd43023093677ad3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3555/108270/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3555/108270/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-22_10.18.24/lib.mongooseim.logs/run.2022-02-22_10.18.24/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/108294) / small_tests / f1202cfa3817fa61e70d0c951dfe67aa6ecbeede
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3555/108294/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3555/108294/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-22_10.22.22/lib.mongooseim.logs/run.2022-02-22_10.22.23/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/108367) / small_tests / 2ba28b5eae02752ce3db42cee340d111f45f2266
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3555/108367/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3555/108367/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-22_11.48.27/lib.mongooseim.logs/run.2022-02-22_11.48.27/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/108389) / small_tests / 634c0cc5e4801ba6925eb4bf55559e57a51986dd
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3555/108389/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3555/108389/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-22_12.03.57/lib.mongooseim.logs/run.2022-02-22_12.03.58/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/108408) / small_tests / 7b99d7bc12ff80e9c56a66c15028308e25590ab4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3555/108408/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3555/108408/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-22_12.43.39/lib.mongooseim.logs/run.2022-02-22_12.43.39/suite.log.html)

File: src/inbox/mod_inbox.erl [hunk 1]
```diff
@@ -536,13 +536,13 @@ config_metrics(HostType) ->
 
 muclight_dep(List) ->
     case lists:member(muclight, List) of
-        true -> [{mod_muc_light, hard}];
+        true -> [{mod_muc_light, [], hard}];
```
Comment (gustawlippa) [inline src/inbox/mod_inbox.erl:539]: From what I understand - this place, and all these similar ones could be a `#{}` already?

Comment (chrzaszcz) [reply src/inbox/mod_inbox.erl:539]: If a module still has its options in a list, it should be `[]`.

**Pull request merged**